src/lib/device/physical_dev.cpp

```diff
@@ -486,232 +217,246 @@ void PhysicalDev::sync_readv(iovec* iov, int iovcnt, uint32_t size, uint64_t off
+        throw std::system_error(e.code(), fmt::format("dev_name: {}: {}", m_devname, e.what()));
     }
 }
 
 void PhysicalDev::sync_write_zero(uint64_t size, uint64_t offset) {
     m_drive_iface->sync_write_zero(m_iodev.get(), size, offset);
 }
 
-void PhysicalDev::attach_chunk(PhysicalDevChunk* chunk, PhysicalDevChunk* after) {
-    if (after) {
-        chunk->set_next_chunk(after->next_chunk_mutable());
-        chunk->set_prev_chunk(after);
-
-        auto* next = after->next_chunk_mutable();
-        if (next) next->set_prev_chunk(chunk);
-        after->set_next_chunk(chunk);
-    } else {
-        HS_DBG_ASSERT_EQ(first_chunk_id(), INVALID_CHUNK_ID);
-        m_info_blk.first_chunk_id = chunk->chunk_id();
-    }
+void PhysicalDev::submit_batch() { m_drive_iface->submit_batch(); }
+
+//////////////////////////// Chunk Creation/Load related methods /////////////////////////////////////////
+void PhysicalDev::format_chunks() {
+    m_chunk_info_slots = std::make_unique< sisl::Bitset >(hs_super_blk::chunk_info_bitmap_size(m_dev_info));
+    auto bitmap_mem = m_chunk_info_slots->serialize(m_pdev_info.dev_attr.align_size);
+    write_super_block(bitmap_mem->bytes, bitmap_mem->size, hs_super_blk::chunk_sb_offset());
 }
 
-std::array< uint32_t, 2 > PhysicalDev::merge_free_chunks(PhysicalDevChunk* chunk) {
-    std::array< uint32_t, 2 > freed_ids{INVALID_CHUNK_ID, INVALID_CHUNK_ID};
-    uint32_t nids{0};
+std::vector< shared< Chunk > > PhysicalDev::create_chunks(const std::vector< uint32_t >& chunk_ids, uint32_t vdev_id,
+                                                          uint64_t size) {
+    std::vector< shared< Chunk > > ret_chunks;
+    std::unique_lock lg{m_chunk_op_mtx};
+    auto chunks_remaining = chunk_ids.size();
+    uint32_t cit{0};
 
-    // Check if previous and next chunk are free, if so make it contiguous chunk
-    PhysicalDevChunk* prev_chunk = chunk->prev_chunk_mutable();
-    PhysicalDevChunk* next_chunk = chunk->next_chunk_mutable();
+    try {
+        while (chunks_remaining > 0) {
+            auto b = m_chunk_info_slots->get_next_contiguous_n_reset_bits(0u, chunks_remaining);
+            if (b.nbits == 0) { throw std::out_of_range("System has no room for additional chunk"); }
+
+            auto buf = hs_utils::iobuf_alloc(chunk_info::size * b.nbits, sisl::buftag::superblk,
+                                             m_pdev_info.dev_attr.align_size);
+            auto ptr = buf;
+            for (auto cslot = b.start_bit; cslot < b.start_bit + b.nbits; ++cslot, ++cit, ptr += chunk_info::size) {
+                chunk_info* cinfo = new (ptr) chunk_info();
+                populate_chunk_info(cinfo, vdev_id, size, chunk_ids[cit], cit);
+
+                auto chunk = std::make_shared< Chunk >(this, *cinfo, cslot);
+                ret_chunks.push_back(chunk);
+                get_stream(chunk).m_chunks_map.insert(std::pair{chunk_ids[cit], std::move(chunk)});
```

Comment:
cinfo->~ chunk_info() will be needed even if no memory will be leak, but it will help to align with Line 307.

Reply:
Yes that is correct, added this line.